File: src/LinkedList/LinkedList.java
package LinkedList;

import java.util.NoSuchElementException;

public class LinkedList {
    private Node first;
    private Node last;
    private int size;

    public void addFirst(int item) {
        Node node = new Node(item);

        if (isEmpty())
            first = last = node;
        else {
            node.next = first;
            first = node;
        }

        size++;
    }

    public void addLast(int item) {
        Node node = new Node(item);

        if (isEmpty())
            first = last = node;
        else {
            last.next = node;
            last = node;
        }

        size++;
    }

    public void removeFirst() {
        if (isEmpty())
            throw new NoSuchElementException();

        if (first == last) {
            first = last = null;
        } else {
            Node second = first.next;
            first.next = null;
            first = second;
        }

        size--;
    }

    public void removeLast() {
        if (isEmpty())
            throw new NoSuchElementException();

        if (first == last) {
            first = last = null;
        } else {
            Node previous = getPrevious(last);
            last = previous;
            last.next = null;
        }

        size--;
    }

    public int indexOf(int item) {
        int index = 0;
        Node current = first;

        while (current != null) {
            if (current.value == item)
                return index;
            current = current.next;
            index++;
        }

        return -1;
    }

    public boolean contains(int item) {
        return indexOf(item) != -1;
    }

    public int size() {
        return size;
    }

    public int[] toArray() {
        int[] items = new int[size];
        Node current = first;

        int index = 0;

        while (current != null) {
            items[index++] = current.value;
            current = current.next;
        }

        return items;
    }

    public void reverse() {
        int[] numbers = this.toArray();

        int count = 0;

        while (count < numbers.length) {
            this.removeLast();
            this.addFirst(numbers[count]);
            count++;
        }
    }

    private boolean isEmpty() {
        return first == null;
    }

    private Node getPrevious(Node node) {
        Node current = first;
        while (current != null) {
            if (current.next == node) return current;
            current = current.next;
        }
        return null;
    }

    private class Node {
        private int value;
        private Node next;

        public Node(int value) {
            this.value = value;
        }
    }
}

File: linked-list/linkedlist/linked_list.go
package linkedlist

import (
	"errors"
	"fmt"
	"reflect"
)

var (
	ErrNoSuchElement = errors.New("no such element")
	ErrOutOfRange    = errors.New("out of range")
)

type linkedList[T any] struct {
	first *node[T]
	last  *node[T]
	size  int
}

func NewLinkedList[T any]() *linkedList[T] {
	return &linkedList[T]{}
}

func (l *linkedList[T]) String() (out string) {
	node := l.first

	for node != nil {
		if node == l.first && node == l.last {
			out += fmt.Sprintf("first-last:%v->", node.val)
		} else if node == l.first {
			out += fmt.Sprintf("first:%v->", node.val)
		} else if node == l.last {
			out += fmt.Sprintf("last:%v->", node.val)
		} else {
			out += fmt.Sprintf("%v->", node.val)
		}
		node = node.next
	}

	return
}

func (l *linkedList[T]) AddFirst(v T) {
	node := newNode(v, nil)

	l.size++

	if l.isEmpty() {
		l.first, l.last = node, node
		return
	}

	node.next = l.first
	l.first = node
}

func (l *linkedList[T]) AddLast(v T) {
	node := newNode(v, nil)

	l.size++

	if l.isEmpty() {
		l.first, l.last = node, node
		return
	}

	l.last.next = node
	l.last = node
}

func (l *linkedList[T]) DeleteFirst() (err error) {
	if l.isEmpty() {
		err = ErrNoSuchElement
		return
	}

	temp := l.first
	l.first = l.first.next
	temp.next = nil
	l.size--
	return
}

func (l *linkedList[T]) DeleteLast() (err error) {
	if l.isEmpty() {
		err = ErrNoSuchElement
		return
	}

	l.size--
	if l.first == l.last {
		l.first, l.last = nil, nil
		return
	}

	prev := l.getPrevious(l.last)

	prev.next = nil
	l.last = prev
	return
}

func (l *linkedList[T]) Contains(v T) bool {
	return l.IndexOf(v) != -1
}

func (l *linkedList[T]) IndexOf(v T) (index int) {
	index = -1

	i := 0
	for node := l.first; node != nil; node = node.next {
		if reflect.DeepEqual(node.val, v) {
			index = i
			return
		}

		i++
	}

	return index
}

func (l *linkedList[T]) Size() int {
	return l.size
}

func (l *linkedList[T]) ToSlice() []T {
	results := make([]T, l.Size())

	node := l.first

	for i := 0; i < len(results); i++ {
		results[i] = node.val
		node = node.next
	}

	return results
}

func (l *linkedList[T]) Reverse() {
	if l.isEmpty() {
		return
	}

	prev := l.first
	cur := l.first.next

	for cur != nil {
		next := cur.next
		cur.next = prev
		prev = cur
		cur = next
	}

	l.last = l.first
	l.last.next = nil
	l.first = prev
}

func (l *linkedList[T]) GetKthFromTheEnd(k int) (val T, err error) {
	if l.isEmpty() {
		err = ErrOutOfRange
		return
	}

	a := l.first
	b := l.first

	for i := 0; i < k-1; i++ {
		b = b.next
		if b == nil {
			err = ErrOutOfRange
			return
		}
	}

	for b != l.last {
		a = a.next
		b = b.next
	}

	val = a.val
	return
}

func (l *linkedList[T]) PrintMiddle() {
	if l.isEmpty() {
		fmt.Println(ErrOutOfRange.Error())
		return
	}

	a := l.first
	b := l.first

	for b != l.last && b.next != l.last {
		b = b.next.next
		a = a.next
	}

	if b == l.last {
		fmt.Println(a.val)
	} else {
		fmt.Println(a.val, ",", a.next.val)
	}
}

func (l *linkedList[T]) isEmpty() bool {
	return l.first == nil
}

func (l *linkedList[T]) getPrevious(n *node[T]) *node[T] {
	if l.isEmpty() || l.first == l.last || n == l.first {
		return nil
	}

	prev := l.first

	for prev.next != n {
		prev = prev.next
	}

	return prev
}
